1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,229 @@
+
+## Motivation
+
+In PowerShell by default, script processing continues when non-terminating errors occur. This is a
+benefit when expecting non-terminating errors in normal execution such as non-responsive computers from
+a list. This default behavior is controlled with the preference variable
+`$ErrorActionPreference` default of `Continue`.
+
+During development and debugging, often customers prefer that script execution stops when a
+non-terminating error occurs. PowerShell currently supports customers with this ability by setting
+the preference variable in the script.
+
+```Powershell
+$ErrorActionPreference = 'Stop'
+```
+
+Native commands return an exit code to the calling application which will be zero for success or
+non-zero for failure. However, native commands currently do not participate in the PowerShell error
+stream. Customers working with native commands in their scripts will need to check the execution
+status after each call using a helper function similar to below:
+
+```Powershell
+if( ! $? )
+{
+    exit $lastexitcode;
+}
+ ```
+
+Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
+support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
+stale.
+
+To support a similar style of error handling with native commands, this specification proposes
+converting native command errors into PowerShell errors similar to the `sh`-compatible
+`set -e` mode.
+
+ The specification and alternative proposals are based on the
+ [Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
+ committee review of the associated
+ [pull request](https://github.com/PowerShell/PowerShell/pull/3523), and
+ [implementation plan](https://github.com/PowerShell/PowerShell-RFC/pull/88#issuecomment-613653678)
+
+## Specification
+
+This RFC proposes including native commands in the error handling framework when the feature is
+enabled by adding an error to the error stream if the exit code of a native command is non-zero.
+
+The `$PSNativeCommandErrorAction` preference variable will implement a version of the
+`$ErrorActionPreference` variable for native commands.
+
+- The value will default to `Ignore` for compatibility with existing behavior.
+- For non-zero exit codes and except for the value 'Ignore', an `ErrorRecord` will be added to
+  `$Error` that wraps the exit code and the command executed that returned the exit code.
+- Initially, only the existing values of `$ErrorActionPreference` will be supported.
+- The set of values may be extended later to include `MatchErrorActionPreference`, which should
+  apply the `$ErrorActionPreference` setting to native commands.
+- A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference`
+  values, where `MatchErrorActionPreference` is converted to the current value of
+  `$ErrorActionPreference`
+
+Valid values for `$ErrorActionPreference`
```

Comment:
At this point I'm somewhat confused:

- We proposed a new preference variable and talked about one possible setting, which relays it to `$ErrorActionPreference`
- We also said the two variables should be convertible from one to the other, implying that `$PSNativeCommandErrorPreference` has one more member than `$ErrorActionPreference`
- But now we've stopped discussing the first preference variable and are describing `$ErrorActionPreference`, which already exists today with a known set of behaviours.

Reply:
Cleaned up section to address.